# Importing

In [14]:
# Import Libraries
import pandas as pd
import numpy as np
import os

In [15]:
# Import Orders Data

path = r'C:\Users\mkock\Desktop\Data Analytics\Data Immersion\Achievement 4\22-03-2022 Instacart Basket Analysis'
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False )

In [16]:
# Import Products Data

df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

# Mixed-Type Data

In [17]:
# Create a Dataframe

df_test = pd.DataFrame()

In [18]:
# Create a mixed type column

df_test['mix'] = ['a', 'b', 1, True]

In [19]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [20]:
# Check for mixed types

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [21]:
# Converting to String

df_test['mix'] = df_test['mix'].astype('str')

# Missing Values

In [22]:
# Finding Missing Values

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [23]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [24]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [25]:
# Creating Clean Database

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [26]:
df_prods.shape

(49693, 5)

In [27]:
df_prods_clean.shape

(49677, 5)

# Duplicates

In [28]:
# Finding duplicates

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [29]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [30]:
# Dropping duplicates

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [31]:
df_prods_clean_no_dups.shape

(49672, 5)

# Task: Question 2

In [32]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

The max value of prices is increadibly high compared to the mean and other statistical values and is most likely a mistake

# Task: Question 3

In [33]:
# Cecking for mixed-types in df_ords

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col+' has mixed-type data')
  else:
    print (col + ' data type is uniform')

order_id data type is uniform
user_id data type is uniform
eval_set data type is uniform
order_number data type is uniform
order_dow data type is uniform
order_hour_of_day data type is uniform
days_since_prior_order data type is uniform


# Task: Question 5

In [34]:
# Finding Missing Values

df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There are 206209 missing values from the 'days_since_prior_order' column. This is most likely due to some customers only placing one order, which leads to missing values in this column

# Task: Question 6

I cannot use the typical methods, like removing the data, that will cause the loss of a lot of lines, and the inforamtion the missing values provide is still important. I also cannot imputate, as it will cause too much disruption to the statistical values, and we can still use the information regarding the missing values. The best option would be to create a new variable to flag the missing values as a customer's first purchase.

In [35]:
# Creating new Dataframe for clean data

df_ords_clean = df_ords

df_ords_clean['days_since_prior_order'].fillna(0, inplace = True)

In [36]:
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     0.0  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

# Task: Question 7

In [37]:
# Checking for duplicates

df_ords_dups = df_ords_clean[df_ords_clean.duplicated()]

In [38]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

No duplicates Found

# Task: Question 8

No Duplicates Found

# Task: Question 9

In [40]:
# Exporting clean dataframes

df_ords_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_cleaned.csv'))

df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_cleaned.csv'))